Simplify nearest neighbor rescale

Kaltxi opened the pull request:
Simplification of nearest neighbor rescaling algorithm which improves minimap regeneration performance for #1605.

After initial attempt at optimization I noticed that the the two inner loops that go over the src image region corresponding to the dest image pixel look like a weighted sum, but in reality samping only a single pixel - the first one. This effectively leads to just taking this pixel (albeit with some additional floating summation and division, which as far as I can see should lead to a constant factor of 1). If we modify this algorithm to actually sample each pixel in the region - we get an averaged blurry image. By the name of the function - we seek a *sharp* rescale, so averaging seems incorrect. Original behavior looks like nearest neighbor interpolation, so I simplfied it to just take the pixel without any additional computations.

This greatly speeds up the scaling function and yields *almost* the same image. The slight differences in some pixels' brightness I attribute to the floating operations error accumulation.

stevecotton commented on src/sdl/utils.cpp:
```diff
@@ -395,86 +395,59 @@ surface scale_surface_legacy(const surface &surf, int w, int h)
+		// In case format of the src surface is not the same as the default format for dst
+		// convert all pixels to the correct format
+		if(surf->format != dst->format) {
+			PLAIN_LOG << "Shite";
```
Forgotten debug statement?

Kaltxi replied:
@stevecotton Yeah, that's what I get for postponing stuff until late at night :(

By the way, how do I handle review commits? Do I add just new commits or maybe amend my commit and rebase force-push if it's a small thing? Is the pull request merged as is or with a squash?

stevecotton replied:
Whether it gets squashed or not depends on whether the person merging it remembers to change the setting in the drop-down menu, and on what they did to the last PR that they merged. This sometimes results in posting "oops ...." comments in the IRC/Discord channel.

So rebase and force-push is slightly preferred.

gfgtdf commented:
It would be good if the commit message contains your discoveries you mentioned in the other post, (that the previous code always reads the same pixels instead of taking an average and that actually taking the average results in a very blurry picture. )

Kaltxi commented:
Wait, the commit message, or the pr message?

gfgtdf commented:
i meant commit message, since that's what people will look at when they'll try to understand the commit in 5 years in case they find it, but if we squashmerge the pr anyways it doesn't matter since we can pick the prs message as commit message .

gfgtdf commented on src/sdl/utils.cpp:
```diff
@@ -395,86 +395,58 @@ surface scale_surface_legacy(const surface &surf, int w, int h)
+		// convert all pixels to the correct format
+		if(surf->format != dst->format) {
+			const int ok
+				= SDL_ConvertPixels(w, h, surf->format->format, src_pixels, 0, dst->format->format, dst_pixels, 0);
```
Is this tested? It looks like it would just copy pixels from src_pixels to dst_pixels, without doing the scaling at all (and basicially overwriting the data in dst_pixels).

Kaltxi replied:
@gfgtdf I did test it, but not well enough, I was expecting image corruption, which didn't happen, but this has an error indeed, it should just have src_pixels in both places, thanks for catching it. I I'll look at the unit testing and will try to add proper tests to make sure.

Kaltxi replied:
@gfgtdf I added the unit tests, but the format verification one didn't work, all surfaces we the same format, even if constructed manually. Turns out `surface` maintains a single format under the hood via `surface::make_neutral` or simply contructing the surface with neutral format. I removed the conversion, since it wasn't doing anything (neither was it reverting the scaling btw, which is somewhat surprising).

Kaltxi commented:
While fixing the pixel format comment, I added unit tests. Not sure if it makes sense to pin the implemntation like this, so I can revert them if needed.

During this process I stumbled against a problem. I used `add_source_file` script to add the test source, which seems to be correct action, since multiple build systems are supported. But failed to do so by specifying the full relative path `py -3 add_source_file src/tests/test_sdl.cpp tests` and worked only with `py -3 add_source_file test_sdl.cpp tests`. This makes sense as we are using source group, but after this CMake failed to build, and in actual source list the path was incorrect, so I had to fix it manually. Did I do something wrong, or is there a bug?

Kaltxi commented:
Now there seem to be problems with building on mac (and linux?) due to possibly incorrect project files. How to register a new file properly? I am on windows at the moment.

soliton- commented:
`add_source_file` is your best bet. Otherwise someone on macos needs to update the project file.

https://github.com/wesnoth/wesnoth/blob/34c8753878a3bdb8173aa6b9a62a3fa40b25d721/add_source_file#L5-L6

Looks like this is a known issue. I suppose the implication is that one needs to manually add it to the tests group.

You could try adding it here:
https://github.com/wesnoth/wesnoth/blob/063eefdc599b429dceb7b31abc92958d3f4bdadb/projectfiles/Xcode/The%20Battle%20for%20Wesnoth.xcodeproj/project.pbxproj#L4779-L4813

Kaltxi commented:
Added the file to tests group and then additionally, looking at the way other files are registered, removed it from src group. If this doesn't fix it, I'm not sure what else could be done with xcode... CI fails on launching the `unit_tests` binary (it couldn't find it), but build step seems to finish without errors.

Pentarctagon commented:
Yeah, that's a known issue - for some reason xcodebuild will exit with a successful return code even when the build actually failed (and I assume the reason for #8256).

soliton- commented:
Did this perhaps happen because you first added the file without specifying the tests target? And then subsequently trying to add it again to the tests target then ran into that issue.

If so then perhaps it would make sense to require a target argument instead of defaulting to wesnoth to try to avoid this.

Another idea would be to remove the file from the wesnoth target (with according log messages) if it's to be added in the tests target since I don't think it makes sense to be in both.

Kaltxi commented:
No, not really, I ran add_source_file script with `tests` target only, but it failed not only to add it to the correct group for xcode but also to specify the correct path for CMake. Then I manually tried to fix it, and it seems I messed up the code one. Now I understand how to fix it correctly, but it's still strange it doesn't work properly...

Kaltxi commented:
Sorry if nagging, not sure how the PR review/merge should go at this point, so could you take another look @stevecotton @gfgtdf?

Pentarctagon commented:
Confirmed on Discord that it's good to merge - thanks for the PR!
